core/src/main/java/org/infinispan/expiration/impl/TxClusterExpirationManager.java

```diff
@@ -12,11 +19,46 @@
 
    @Inject
    protected KeyPartitioner partitioner;
+   @Inject @ComponentName(KnownComponentNames.PERSISTENCE_EXECUTOR)
+   protected ExecutorService blockingExecutor;
+
+   private boolean optimisticTransaction;
+
+   @Override
+   public void start() {
+      super.start();
+
+      optimisticTransaction = configuration.transaction().lockingMode() == LockingMode.OPTIMISTIC;
+   }
 
    @Override
    CompletableFuture<Boolean> handleMaxIdleExpireEntry(K key, V value, long maxIdle, boolean skipLocking) {
       // We cannot remove an entry in a tx, due to possible dead lock - thus we just return
       // if the entry is expired or not
-      return retrieveLastAccess(key, value, partitioner.getSegment(key)).thenApply(hasMaxIdle -> hasMaxIdle == null);
+      return retrieveLastAccess(key, value, partitioner.getSegment(key)).thenApply(Objects::isNull);
+   }
+
+   @Override
+   CompletableFuture<Void> removeLifespan(AdvancedCache<K, V> cacheToUse, K key, V value, long lifespan) {
+      // Transactional is still blocking - to be removed later
+      return CompletableFuture.supplyAsync(() -> super.removeLifespan(cacheToUse, key, value, lifespan), blockingExecutor)
+            .thenCompose(Function.identity());
+   }
+
+   @Override
+   CompletableFuture<Boolean> removeMaxIdle(AdvancedCache<K, V> cacheToUse, K key, V value) {
+      // Transactional is still blocking - to be removed later
```

Comment:
I'm not sure if this is still required. 
`CacheImpl.performRemoveExpiredCommand() `suspends the current transaction and creates a new implicit transaction.
The implicit transaction is then completed after the CompletableFuture is done.

Reply:
It definitely improves performance of reaper, so something is still blocking in there.

Reply:
Forking immediately to another thread could speed up the expiration even if the transaction is committed asynchronously, because starting a transaction is expensive. The effect should be much smaller than on 9.4.x though, assuming 9.4.x was committing the transaction on the same thread.